### ISSUE TITLE
some tests fail on FreeBSD/Hurd due to lack of fsync()

### ISSUE THREAD

**Issue by jelmer** [dulwich/dulwich#541]
ERROR: test_simple (dulwich.tests.test_porcelain.PushTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "dulwich/tests/test_porcelain.py", line 478, in test_simple
    self.assertEqual(r_clone[b'HEAD'].id, self.repo[refs_path].id)
  File "dulwich/repo.py", line 456, in __getitem__
    return self.object_store[self.refs[name]]
  File "dulwich/object_store.py", line 116, in __getitem__
    type_num, uncomp = self.get_raw(sha)
  File "dulwich/object_store.py", line 378, in get_raw
    raise KeyError(hexsha)
KeyError: '1a588fa84ea70b3ca8e9d40d6a34c19dfb5a3e07'

======================================================================
ERROR: test_fetch_from_dulwich_issue_88_alternative (dulwich.tests.compat.test_web.DumbWebTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "dulwich/tests/compat/server_utils.py", line 269, in test_fetch_from_dulwich_issue_88_alternative
    b'02a14da1fc1fc13389bbf32f0af7d8899f2b2323').type_name)
  File "dulwich/repo.py", line 337, in get_object
    return self.object_store[sha]
  File "dulwich/object_store.py", line 116, in __getitem__
    type_num, uncomp = self.get_raw(sha)
  File "dulwich/object_store.py", line 378, in get_raw
    raise KeyError(hexsha)
KeyError: '02a14da1fc1fc13389bbf32f0af7d8899f2b2323'

(https://buildd.debian.org/status/fetch.php?pkg=dulwich&arch=kfreebsd-i386&ver=0.12.0-1&stamp=1452746352)

**Comment by jelmer** [comment]
This still happens, it probably has a different cause.

**Comment by sbraz** [comment]
I can see this on Linux too. It seems to fail every other time I call it.
```
PYTHONPATH=. python2 -m unittest "dulwich.tests.test_porcelain.PushTests.test_simple"
E
======================================================================
ERROR: test_simple (dulwich.tests.test_porcelain.PushTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "dulwich/tests/test_porcelain.py", line 682, in test_simple
    self.assertEqual(r_clone[b'HEAD'].id, self.repo[refs_path].id)
  File "dulwich/repo.py", line 476, in __getitem__
    return self.object_store[self.refs[name]]
  File "dulwich/object_store.py", line 118, in __getitem__
    type_num, uncomp = self.get_raw(sha)
  File "dulwich/object_store.py", line 423, in get_raw
    raise KeyError(hexsha)
KeyError: 'fdb9f14fd6c66c88eb64d49049248084f6815615'

----------------------------------------------------------------------
Ran 1 test in 0.019s

FAILED (errors=1)
```

**Comment by jelmer** [comment]
Interesting. What filesystem is your /tmp on?

**Comment by sbraz** [comment]
It is a tmpfs. Since you asked, I tried mounting an ext4 partition there but the tests still failed sometimes.
I noticed that if I run the tests with strace, making them a bit slower, they never fail. So I suspect a race condition (but you probably guessed that already).

**Comment by jelmer** [comment]
Yes, this is indeed a race condition. I was thinking it could be avoided by using data=ordered (as on ext4 by default), so it's interesting to hear this can be reproduced with ext4 too.

**Comment by stefanor** [comment]
We're seeing it regularly in Ubuntu autopkgtests on s390x too: http://autopkgtest.ubuntu.com/packages/d/dulwich/bionic/s390x

**Comment by jelmer** [comment]
This is fixed in dulwich 0.19.11
